### DISCUSSION TITLE
Test order

### DISCUSSION THREAD

**Discussion by bigcurl** [bigcurl/ffi-cuda-toolkit-gem-generator#9] [General]
Hi @s96Abrar,

I had thought about the order of which the tests should be written. Here is a sorted list. You can create a new file in the test folder and name it like the modules name. See the existings files for reference. There is also a list of modules that do not need tests (mainly the graphics apis):

6. Modules
    - [x] 6.2. Error Handling
    - [x] 6.3. Initialization
    - [x] 6.4. Version Management
    - [x] 6.16. Event Management
    - [x] 6.5. Device Management
    - [x] 6.7. Primary Context Management
    - [x] 6.11. Memory Management
    - [x] 6.10. Module Management
    - [ ] 6.19. Execution Control
    - [x] 6.15. Stream Management
    - [ ] 6.21. Graph Management
    - [ ] 6.13. Stream Ordered Memory Allocator
    - [ ] 6.22. Occupancy
    - [ ] 6.31. Profiler Control
    - [ ] 6.14. Unified Addressing
    - [x] 6.8. Context Management
    - [ ] 6.12. Virtual Memory Management
    
These ones are not needed as I have no use for the graphics capabilities yet:
    6.17. External Resource Interoperability
    6.18. Stream memory operations
    6.25. Texture Object Management
    6.26. Surface Object Management
    6.28. Graphics Interoperability
    6.33. VDPAU Interoperability
    6.34. EGL Interoperability
    6.32. OpenGL Interoperability
    6.32.1. OpenGL Interoperability [DEPRECATED]
    6.23. Texture Reference Management [DEPRECATED]
    6.24. Surface Reference Management [DEPRECATED]
    6.27. Peer Context Memory Access
    6.29. Driver Entry Point Access
    6.6. Device Management [DEPRECATED]
    6.9. Context Management [DEPRECATED]
    6.20. Execution Control [DEPRECATED]
    6.30. Profiler Control [DEPRECATED]

**Comment by bigcurl** [comment]
Hi @s96Abrar,

just wanted to check in on you, if everything is alright so far. I can only tell you that you are writing great ruby code even though it was new for you. I was sure that the language should be no problem for you. Keep on with the good work.

**Reply by s96Abrar** [reply]
Thanks. I haven't saw the message before just found it now. 

Everything is going well and good.

Just wanted to let you know that as you see I am going very slowly writing those tests, understanding the documentation and the correct values for the function parameter is a bit hard as I have not much concept on graphics related stuff like how a memory stored for a graphics object. Was having a little bit confusing while writing the memory management test (Not yet covered all methods).

Hope will get to a good speed writing the test cases.

Also I saw that running rake only checks driver_api for version 11.5 not checking the 11.4.2

**Reply by bigcurl** [reply]
> Also I saw that running rake only checks driver_api for version 11.5 not checking the 11.4.2

Yes you are right, it is only checking 11.5. To check the tests for 11.4, a copy of the tests is needed. I will do that after the version for 11.5 is ready as the tests are almost the same.

> Was having a little bit confusing while writing the memory management test (Not yet covered all methods).

Not all tests are needed. All methods related to graphics work are not needed. If you commit your current version of the tests I will add a list of the needed methods at the top of the file and if you mark which ones are confusing for you, I might be able to help a bit.

Let me know if this works for you?

**Reply by s96Abrar** [reply]
Sorry I again forgot to reply. Yeah that will be great.

**Reply by bigcurl** [reply]
ok, than please wait until I commit the reorganization of the repo to a gem only repo. I will also mark the needed methods and than let you know. If you want you can continue in the meantime but I will put everything back into the main branch, so maybe a little bit of merging is required.

**Reply by s96Abrar** [reply]
OK no problem.

**Comment by bigcurl** [comment]
- The restruction of the repo is done. I removed the parser part and now it is only the gem.
- Also all test files are now updated with needed methods and not needed methods. 
- I also moved some files in the ordered modules list to not needed.

**Comment by s96Abrar** [comment]
When you be free have a look at the latest memory test. At `test_cu_mem_cpy_h_to_d` and `test_cu_mem_cpy_h_to_d_async` having trouble. Function `cuMemcpyHtoD_v2` can not copy the host memory data to device because of invalid value provided in to the function.

**Comment by bigcurl** [comment]
Hi @s96Abrar,

took me a while to correct the failing memory tests. It came down to a wrong definition in the wrapper file. The type `cuDevicePtr` was an `int` but it is a `long long`. It is an `int` in the documentation but a function that evaluates the target plattform in the header file. See my changes in commit be306e2358be86e5f08dd43be80fd5adbf99b4f4 for what i have done.

**Reply by s96Abrar** [reply]
Awesome. I really never understood how to store the memory address returned as CUdeviceptr in ruby.

**Reply by s96Abrar** [reply]
[Letting you know] At the module management seems like we do need to test some external C code written as ptx, cubin, fatbin but not sure.

From the Cuda samples as I saw module management like `cuModuleLoadData` function load some external files. (at 0_Simple/matrixMulDrv program).

I am skipping that for now.

**Reply by bigcurl** [reply]
Try to implement the test for function `cuModuleLoadFatBinary` from the `Module Management` module. You can find a `fatbin` file after running `make` in the folder `NVIDIA_CUDA-11.5_Samples/0_Simple/vectorAddDrv`. For me it generates the file `vectorAdd_kernel64.fatbin`. Upload it into the repo as well. 
Loading a cuda kernel is an essential operation, I think even for the test implementations that come after the Module Management tests.

**Comment by bigcurl** [comment]
Hi @s96Abrar,

just wanted to know if you are still on the project? Otherwise you can create a project on fiverr and I will send the money. Project is bigger than anticipated and you already helped me a lot.

**Reply by s96Abrar** [reply]
Hi, yeah I do like to stay on the project but dew to some exams of mine recently and I was not able to manage time quite well to focus on the project yet. Sorry that I should have let you know before. I will try to start very recently within a week.

**Reply by bigcurl** [reply]
Well than good luck with your exams. What and where are you studying?

**Reply by s96Abrar** [reply]
Thanks. 

My major is Software Engineering at Noakhali Science and Technology University in Bangladesh.

**Reply by bigcurl** [reply]
Hi @s96Abrar,

hope your exams went fine. Nice to see a new commit into the project. Do not forget to add your name to the GEM file as author. Maybe it gets you a job at NVIDIA (The Jetson Nano is available in Bangladesh by the way). Cheap way to get your hands on an NVIDIA GPU and play more with CUDA.

**Reply by s96Abrar** [reply]
Hi,
Sorry for late reply. I am not fully doing the work yet. I will try to get one.

**Comment by s96Abrar** [comment]
The remaining of the modules that I could not complete because the documentation not seems to clear to me and for some reason it looks tough to call those API functions from ruby.

 - `Execution Control`
 - `Graph Management`
 - `Stream Ordered Memory Allocator`
 - `Occupancy`
 - `Unified Addressing`
 
 Sorry that I have been holding the project in the current stage for so long.

**Reply by bigcurl** [reply]
No Problem @s96Abrar, I'll see what I can do about it. I let you know when I release the project officially (but will take some more time). Please open the project on Fiverr for your payment.
